localstack/services/apigateway/integration.py

```diff
@@ -17,9 +21,86 @@
+    WHEN_NO_MATCH = "WHEN_NO_MATCH"
+    WHEN_NO_TEMPLATES = "WHEN_NO_TEMPLATES"
+    NEVER = "NEVER"
+
+
+class MappingTemplates:
+    """
+    API Gateway uses mapping templates to transform incoming requests before they are sent to the
+    integration back end. With API Gateway, you can define one mapping template for each possible
+    content type. The content type selection is based on the Content-Type header of the incoming
+    request. If no content type is specified in the request, API Gateway uses an application/json
+    mapping template. By default, mapping templates are configured to simply pass through the
+    request input. Mapping templates use Apache Velocity to generate a request to your back end.
+    """
+
+    passthrough_behavior: PassthroughBehavior
+
+    class UnsupportedMediaType(Exception):
+        pass
+
+    def __init__(self, passthrough_behaviour: str):
+        self.passthrough_behavior = self.passthrough_behavior(passthrough_behaviour)
+
+    def request_body_passthrough(self, request_template):
+        """
+        Specifies how the method request body of an unmapped content type will be passed through
+        the integration request to the back end without transformation.
+        A content type is unmapped if no mapping template is defined in the integration or the
+        content type does not match any of the mapped content types, as specified in requestTemplates
+        """
+        if not request_template and self.passthrough_behavior in {
+            PassthroughBehavior.NEVER,
+            PassthroughBehavior.WHEN_NO_TEMPLATES,
+        }:
+            raise MappingTemplates.UnsupportedMediaType()
+
+    @staticmethod
+    def passthrough_behavior(passthrough_behaviour: str):
+        return getattr(PassthroughBehavior, passthrough_behaviour, None)
+
+
 class BackendIntegration(ABC):
     """Abstract base class representing a backend integration"""
 
+    def __init__(self):
+        self.request_templates = RequestTemplates()
+        self.response_templates = ResponseTemplates()
+
+    @classmethod
+    def _create_response(cls, status_code, headers, data=""):
+        response = Response()
+        response.status_code = status_code
+        response.headers = headers
+        response._content = data
+        return response
+
+    @classmethod
+    def apply_response_parameters(
+        cls, invocation_context: ApiInvocationContext, response: Response
+    ):
+        integration = invocation_context.integration
+        integration_responses = integration.get("integrationResponses") or {}
+        if not integration_responses:
+            return response
+        entries = list(integration_responses.keys())
+        return_code = str(response.status_code)
+        if return_code not in entries:
+            if len(entries) > 1:
+                LOG.info("Found multiple integration response status codes: %s", entries)
+                return response
+            return_code = entries[0]
```

Comment:
I'll try to, this code is just a copy of the existing code and will stay like that until we migrate all the integrations into classes. Said that, this is my interpretation of the code and based on some documentation: 

1. Non-Proxy lambdas don't need to return a status code, in case of runtime errors, this is what the [documentation](https://docs.aws.amazon.com/lambda/latest/dg/troubleshooting-execution.html) states.
```
When your function code or the Lambda runtime return an error, the status code in the response from Lambda is 200 OK
```. So, if the integration response has only one configured status code the response will be whatever the lambda returns (errors or non errors) with the status code configured not 200 (for errors).

2. `entries[0]` is technically not the right pick, because integration response needs a default configuration, coming in a next PR.

Non-proxy (non `AWS_PROXY`), response_template status code trump whatever the lambda returns.

3. For `AWS_PROXY` lambda integrations, whatever is defined in the lambda `statusCode` property, is returned to the client.

The code is not perfect for all use cases (`AWS` and `AWS_PROXY`) but it's something in need to improve, not completely wrong tho. I'll leave as is, and keep iterating.

Reply:
Makes sense, thanks for clarifying. Actually, my example from above was incorrect - the `len(entries) > 1` guards against us selecting a random return code if multiple options are in place. Let's leave as-is and then iterate for `AWS` vs `AWS_PROXY` integrations.. 👍